CI run against latest deps is failing

See https://github.com/matrix-org/synapse/actions/runs/3125758020

Complement flake, plus mypy:

```
synapse/http/server.py:751: error: Incompatible types in assignment (expression has type "Callable[[Any], bytes]", variable has type "Callable[[object], bytes]")  [assignment]
Found 1 error in 1 file (checked 734 source files)
```

This is due to https://github.com/matrix-org/python-canonicaljson/pull/52

Fix by either marking `encoder: Callable[[Any], bytes]`; by making `_encode_json_bytes` take `object`; or with a `#type: ignore`.

--- a/changelog.d/13905.misc
+++ b/changelog.d/13905.misc
@@ -1,0 +1,1 @@
+Fix mypy errors with canonicaljson 1.6.3.

--- a/synapse/http/server.py
+++ b/synapse/http/server.py
@@ -705,7 +705,7 @@ class _ByteProducer:
         self._request = None
 
 
-def _encode_json_bytes(json_object: Any) -> bytes:
+def _encode_json_bytes(json_object: object) -> bytes:
     """
     Encode an object into JSON. Returns an iterator of bytes.
     """
@@ -746,7 +746,7 @@ def respond_with_json(
         return None
 
     if canonical_json:
-        encoder = encode_canonical_json
+        encoder: Callable[[object], bytes] = encode_canonical_json
     else:
         encoder = _encode_json_bytes
 

